README.md
# numista-api-sdk
API Documentation: [https://en.numista.com/api/doc/index.php](https://en.numista.com/api/doc/index.php)

## Code Coverage
[![GitHub Super-Linter](https://github.com/namachieli/numista-api-sdk/workflows/Lint%20Code%20Base/badge.svg)](https://github.com/marketplace/actions/super-linter)

## Installation
### Setup Virtual Environment
A Virtual Environment (venv) is a local copy of a runtime environment that is managed in a 'sandbox' that is seperate from your global environment. For example, your global python install is Python 3.8, but you need to run something in Python 2.7. You can make a virtual environment for Python 2.7, and then execute code from within that venv.

[Read more about Virtual Environments](https://www.geeksforgeeks.org/python-virtual-environment/?ref=rp)

```bash
mkdir ~/inventory && cd ~/inventory
python3 -m venv env
source ~/inventory/env/bin/activate
```

To exit the venv:
```bash
deactivate
```
### pypi.org
This will be added to PYPI soon

### Direct with PIP and SSH
Requires proper SSH Key access to be setup for your client.
```bash
pip install git+ssh://git@github.com/namachieli/numista-api-sdk
pip show numista
```
### HTTPS Clone and Local install
If you do not have an ssh key setup for git, you can clone the repository and install from a local copy.
```bash
mkdir -p ~/git && cd ~/git
git clone https://github.com/namachieli/numista-api-sdk
pip install ~/git/numista-api-sdk
pip show numista
```
## Quickstart
### Import
```python
from numista import Numista
api_key = 'your key'
n = Numista(api_key=api_key)
```
### Query a user
```python
user = n.getUser(user_id=2)
```
**Result**
```python
user
{
    'data':
    {
        'username': 'Xavier',
        'avatar': 'https://en.numista.com/forum/avatars/20002403351640d537b85a.png'
    },
    'http_info':
    {
        'http_status': 200,
        'http_msg': 'Request successful'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [200]>
    }
}
```
## Examples
### Search for a type
```python
escudos = n.searchTypes(q="Escudo", issuer='mexico')
```
**Result**
```python
escudos
{
    'data':
    {
        'count': 20,
        'types':
        [
            {
                'id': 236498,
                'title': 'Medal - Ron Solera Bacardi',
                'category': 'exonumia',
                'issuer':
                {
                    'code': 'mexique',
                    'name': 'Mexico'
                },
                'obverse_thumbnail': 'https://en.numista.com/catalogue/photos/mexique/6062e2f91b56c0.75792711-180.jpg',
                'reverse_thumbnail': 'https://en.numista.com/catalogue/photos/mexique/6062e2f9a76b82.65792937-180.jpg'
            },
            {<truncated>}
        ]
    },
    'http_info':
    {
        'http_status': 200,
        'http_msg': 'Request successful'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [200]>
    }
}
```
### Get a list of Issuers
```python
issuers = n.getIssuers()
```
**Result**
```python
issuers
{
    'data':
    {
        'count': 4239,
        'issuers':
        [
            {
                'code': 'abkhazia',
                'name': 'Abkhazia',
                'wikidata_id': 'Q23334'
            },
            {
                'code': 'afghanistan',
                'name': 'Afghanistan',
                'wikidata_id': 'Q889'
            },
            {<truncated>}
        ]
    },
    'http_info':
    {
        'http_status': 200,
        'http_msg': 'Request successful'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [200]>
    }
}
```
### Get a list of Catalogs/Catalogues
```python
catalogs = n.getCatalogs()  # Alternatively for our UK/EU friends n.getCatalogues()
```
**Result**
```python
catalogs
{
    'data':
    {
        'count': 1587,
        'catalogues':
        [
            {
                'id': 206,
                'code': 'A',
                'title': 'A Checklist of Islamic Coins',
                'author': 'Stephen Album',
                'publisher': 'Self-published',
                'isbn13': '9780615562445'
            },
            {
                'id': 241,
                'code': 'AB',
                'title': 'Catálogo de la moneda medieval castellano-leonesa',
                'author': 'Fernando Álvarez Burgos',
                'publisher': 'Vico-Segarra Editores',
                'isbn13': '9788485711185'
            }, {<truncated>}
        ]
    },
    'http_info':
    {
        'http_status': 200,
        'http_msg': 'Request successful'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [200]>
    }
}
```
### OAuth and list Collections
The generated token is placed under the label `self` in `n.oauthTokens`

```python
n.myTokenGenerate()
{
    'token': '1234abcd5678efgh',
    'user_id': 2,
    'type': 'bearer',
    'scope': 'view_collection,edit_collection',
    'exp_epoch': 1648797646,
    'exp_date': '2022-04-01 00:20:46'
}
result = n.getUserCollections()
```
**Result**
```python
result
{
    'data':
    {
        'count': 5,
        'collections':
        [
            {
                'id': 43709,
                'name': 'Misc - Non-US'
            },
            {
                'id': 43708,
                'name': 'Misc - US'
            },
            {
                'id': 43706,
                'name': 'Type - France'
            },
            {
                'id': 43705,
                'name': 'Type - Mexico'
            },
            {
                'id': 43707,
                'name': 'Type - US'
            }
        ]
    },
    'http_info':
    {
        'http_status': 200,
        'http_msg': 'Request successful'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [200]>
    }
}
```

### Get example schema for adding an item
**note** method would be either `post` for an add, or `patch` for an edit.
```python
schema = n.schemaGenerateBody(operationId="addCollectedItems", method="post")
```
**Result**
```python
schema
{
    'type': 11331,
    'issue': 63444,
    'quantity': 1,
    'grade': 'vf',
    'for_swap': False,
    'private_comment': 'Test with the API',
    'price':
    {
        'value': 76,
        'currency': 'EUR'
    }
}
```
### Add an item to a collection
#### Update schema fields to add a US 1833 Capped Bust Half Dollar
```python
schema['type'] = 10637
schema['issue'] = 73608
schema['grade'] = "au"
del schema['private_comment']
schema['price'] = {"value": 250, "currency": "USD"}
schema['collection'] = 43708

schema
{
    'type': 10637,
    'issue': 73608,
    'quantity': 1,
    'grade': 'au',
    'for_swap': False,
    'price':
        {
            'value': 250,
            'currency': 'USD'
        },
    'collection': 43708
}
```
#### Add the coin
```python
result = n.addCollectedItem(body=schema)
```
**Result**
```python
result
{
    'data':
    {
        'id': 48567296,
        'quantity': 1,
        'type':
        {
            'id': 10637,
            'title': '50 Cents / ½ Dollar &quot;Capped Bust Half Dollar&quot;',
            'category': 'coin',
            'issuer':
            {
                'code': 'etats-unis',
                'name': 'United States'
            }
        },
        'issue':
        {
            'id': 73608,
            'is_dated': True,
            'year': 1833,
            'gregorian_year': 1833,
            'mintage': 5206000
        },
        'for_swap': False,
        'grade': 'au',
        'price':
        {
            'value': 250,
            'currency': 'USD'
        },
        'collection': 43708
    },
    'http_info':
    {
        'http_status': 201,
        'http_msg': 'The requested operation was accepted and successful'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [201]>
    }
}
```
### Get collected items from a collection
```python
result = n.getCollectedItems(collection=43708)
```
**Result**
```python
result
{
    'data':
    {
        'item_count': 1,
        'item_for_swap_count': 1,
        'item_type_count': 1,
        'item_type_for_swap_count': 1,
        'items':
        [
            {
                'id': 48567296,
                'quantity': 1,
                'type':
                {
                    'id': 10637,
                    'title': '50 Cents / ½ Dollar &quot;Capped Bust Half Dollar&quot;',
                    'category': 'coin',
                    'issuer':
                    {
                        'code': 'etats-unis',
                        'name': 'United States'
                    }
                },
                'issue':
                {
                    'id': 73608,
                    'is_dated': True,
                    'year': 1833,
                    'gregorian_year': 1833,
                    'mintage': 5206000
                },
                'for_swap': False,
                'grade': 'au',
                'price':
                {
                    'value': 250,
                    'currency': 'USD'
                },
                'collection':
                {
                    'id': 43708,
                    'name': 'Misc - US'
                }
            }
        ]
    },
    'http_info':
    {
        'http_status': 200,
        'http_msg': 'Request successful'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [200]>
    }
}
```
### Edit collected item
**note** This is a 'patch' operation, so only pass fields you want to edit, with desired new values
```python
body = {"price": {"value": 300, "currency": "USD"}}
result = n.editCollectedItem(item_id=48567296, body=body)
```
**Result**
```python
result
{
    'data':
    {
        'id': 48567296,
        'quantity': 1,
        'type':
        {
            'id': 10637,
            'title': '50 Cents / ½ Dollar &quot;Capped Bust Half Dollar&quot;',
            'category': 'coin',
            'issuer':
            {
                'code': 'etats-unis',
                'name': 'United States'
            }
        },
        'issue':
        {
            'id': 73608,
            'is_dated': True,
            'year': 1833,
            'gregorian_year': 1833,
            'mintage': 5206000
        },
        'for_swap': False,
        'grade': 'au',
        'price':
        {
            'value': 300,
            'currency': 'USD'
        },
        'collection':
        {
            'id': 43708,
            'name': 'Misc - US'
        }
    },
    'http_info':
    {
        'http_status': 200,
        'http_msg': 'Request successful'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [200]>
    }
}
```
### Delete collected item
```python
result = n.deleteCollectedItem(item_id=48567296)
```
**Result**
```python
result
{
    'data':
    {
        'content': ''
    },
    'http_info':
    {
        'http_status': 204,
        'http_msg': 'The item has been deleted'
    },
    'failed': False,
    'extra':
    {
        'requests': <Response [204]>
    }
}
```
## Documentation
Please see the [/docs](/docs) folder for verbose documentation on other methods and capabilities.

## Contribution
Pull requests are always welcome!


CHANGELOG.md
# Changelog

## 0.1.0
### Changes
- None

### Additions
- Initial development
- Methods for API interaction
- Methods for ease of use
- OAuth capabilities for creation of Tokens to interact with User data
- Schema interpreters to help with payload formatting of POST/PATCH operations
- Auto generated code documentation
- Automated Linting (Super Linter)

### Deferred
- Payload schemas are 'Example' versions only. Future: Full schema ingestion parsing
- Body Payload validation is only validating the body exists for POST/PATCH. Future: Validate against ingested Schema and inform of granular errors
- Log level is either INFO (default) or DEBUG (bool). Future: Enable the changing of log levels before and after instantiation.
- Log path is not validated for existence, let alone iswritable(). Future: Path validation for log file location
- API Keys and Tokens are being explicitely obfuscated in logs. Future: Use log filters to automate obfuscation of sensitive data from logs


static/add_to_my_collection.form.script.js
/* This file is for reference purposes, scraped from live on Mar 14 2022 */
function save_collection(add_again) {
    $('.modal .button').attr("disabled", "disabled");
    let collectibleType = $('#collec_form_ct').val();
    let coin = $('#collec_form_type').val();
    let version = $('#collec_form_coin').val();
    $.post('/vous/save_collection.php', {
        'coinId':       coin,
        'version':      version,
        'item':         $('#collec_form_item').val(),
        'quantity':     $('#collec_form_quantity').val(),
        'grade':        $('input[name=collec_form_grade]:checked').val(),
        'value':        $('#collec_form_value').val(),
        'comment':      $('#collec_form_comment').val(),
        'forSwap':      $('input[name=collec_form_swap]:checked').val(),
        'swapComment':  $('#collec_form_swap_comment').val(),
        'section':      $('#collec_form_section').val(),
        'pictures':     $('#collec_form_pictures').val()})
    .done(function(response) {
        if (response.substring(0, 5)==='LOGIN') {
            alert("You were disconnected. Please log in and try again.");
            $('.modal .button').removeAttr("disabled");
        }
        else if (response.substring(0, 3)==='ERR') {
            alert("An error occurred. Please try again.");
            $('.modal .button').removeAttr("disabled");
        }
        else {
            if (version && version!=='') {
                $('#collec_line' + version).html(response);
                version = parseInt(version)
            }
            else {
                $('#collec_line0_' + coin).html(response);
                show_undetermined_line(coin);
                version = null;
            }
            activate_thumbnails();
            refresh_collection_count(coin);
            collec_modal_close();
            if (add_again) collec_modal_new(collectibleType, coin, version);
        }
    })
    .fail(function() {
        alert("An error occurred. Please try again.");
        $('.modal .button').removeAttr("disabled");
    });
}

function refresh_collection_count(coin) {
    if ($('#affichage_qc' + coin).length) {
        let collec_count = 0;
        $('.collec_q.col' + coin).each(function() { collec_count += parseInt($(this).text()); });
        $('#affichage_qc' + coin).text(collec_count);
    }
    if ($('#affichage_qe' + coin).length) {
        let swap_count = 0;
        $('.collec_q.swa' + coin).each(function() { swap_count += parseInt($(this).text()); });
        $('#affichage_qe' + coin).text(swap_count);
        if (swap_count) $('#affichage_qe_cadre' + coin).css('display', 'inline');
    }
    if ($('#affichage_qt' + coin).length) {
        let total_count = 0;
        $('.collec_q.col' + coin).each(function() { total_count += parseInt($(this).text()); });
        $('.collec_q.swa' + coin).each(function() { total_count += parseInt($(this).text()); });
        $('#affichage_qt' + coin).text(total_count);
    }
}

function delete_collection_item(item, version, coin) {
    if(!confirm("Do you wish to remove this item from your collection?")) return;
    $.post('/vous/remove_collection_item.php', {'item':item, 'version':version, 'collectible':coin})
    .done(function(response) {
        if (response.substring(0, 5)==='LOGIN') {
            alert("You were disconnected. Please log in and try again.");
        }
        else if (response.substring(0, 3)==='ERR') {
            alert("An error occurred. Please try again.");
        }
        else {
            if (version) $('#collec_line' + version).html(response);
            else $('#collec_line0_' + coin).html(response);
            refresh_collection_count(coin);
        }
    })
    .fail(function() {
        alert("An error occurred. Please try again.");
    });
}

function show_undetermined_line(collectible_id) {
    $('.undetermined_line' + collectible_id).css('display', '');
    $('.undetermined_line' + collectible_id + ' button').attr('tabindex', '0');
    $('.undetermined_filler' + collectible_id).remove();
}

function collec_modal_close() {
    $('.modal').remove();
}

function htmlEncode(value) {
    return value.replace(/</g,"&lt;").replace(/>/g,"&gt;").replace(/"/g,"&quot;");
}

function collec_modal_new(collectibleType, coinId, selected_version, item, quantity, grade, forSwap, value, comment, swapComment, section, pictures) {
    selected_version = typeof selected_version!=='undefined'? selected_version: null;
    item = typeof item!=='undefined'? item: null;
    quantity = typeof quantity!=='undefined'? quantity: null;
    grade = typeof grade!=='undefined'? grade: null;
    forSwap = typeof forSwap!=='undefined'? forSwap: null;
    value = typeof value!=='undefined'? value: null;
    comment = typeof comment!=='undefined'? comment: null;    
    swapComment = typeof swapComment!=='undefined'? swapComment: null;
            section = (typeof section!=='undefined' && section!=0)? section: 43707;
            pictures = typeof pictures!=='undefined'? pictures: null;
    let modal = document.createElement('section');
    let action = item==null?'create':'edit';
    modal.className = 'modal';
    let modal_title = action==='create'? 'Add to my collection': 'Edit items in my collection';

    let content = '<div class="modal-content" role="dialog" aria-labelledby="modal_title" aria-modal="true"><form action="" method="get">\
        <input type="button" class="close" title="Close" onclick="collec_modal_close();" value="&times" />\
        <header id="modal_title">' + modal_title + '</header>\
        \
        <div id="collec_form">\
        <input type="hidden" name="collec_form_ct" id="collec_form_ct" value="' + collectibleType +'">\
        <input type="hidden" name="collec_form_type" id="collec_form_type" value="' + coinId +'">\
            <input type="hidden" name="collec_form_item" id="collec_form_item" value="' + (item==null?'':item) + '">';
        
    if (typeof coin_data!='undefined') {
        let versions = coin_data['c'+coinId].versions
        content = content + '\
            <div>\
                <label for="collec_form_coin">' + (collectibleType==='coin'?'Coin:':(collectibleType==='banknote'?'Banknote:':'Version:')) + '</label>\
                <span class="form_value">\
                    <select name="collec_form_coin" id="collec_form_coin"';
    if (action==='edit') content = content + ' disabled="disabled"';
    content = content + '>';
    if (versions.length > 1) {
        let selected = selected_version==null? 'selected="selected"': '';
        content = content  + '<option value="" ' + selected + '>Undetermined</option>';
    }
    for (let i = 0; i < versions.length; i++) {
        let version = versions[i];
        let selected = version[0]===selected_version? 'selected="selected"': '';
        content = content  + '<option value="' + version[0] + '" ' + selected + '>' + version[1] + ' ' + version[2] + ' ' + version[3] +  '</option>';
    }
    content = content + '\
                    </select>\
                </span>\
            </div>';
    }

    else {
        content = content + '\
            <input type="hidden" name="collec_form_coin" id="collec_form_coin" value="' + selected_version + '" />';
    }

    content = content + '\
            <div>\
                <span class="form_name">Grade:</span>\
                <span class="form_value">\
                    <input type="radio" name="collec_form_grade" value="ab" id="collec_form_grade_ab"' + (grade==='ab'?' checked="checked"':'') + ' /><label for="collec_form_grade_ab"><abbr title="' + (collectibleType==='coin'?'Good':(collectibleType==='banknote'?'Good':'Good')) + '">' + (collectibleType==='coin'?'G':(collectibleType==='banknote'?'G':'G')) + '</abbr></label>\<input type="radio" name="collec_form_grade" value="b" id="collec_form_grade_b"' + (grade==='b'?' checked="checked"':'') + ' /><label for="collec_form_grade_b"><abbr title="' + (collectibleType==='coin'?'Very good':(collectibleType==='banknote'?'Very good':'Very good')) + '">' + (collectibleType==='coin'?'VG':(collectibleType==='banknote'?'VG':'VG')) + '</abbr></label>\<input type="radio" name="collec_form_grade" value="tb" id="collec_form_grade_tb"' + (grade==='tb'?' checked="checked"':'') + ' /><label for="collec_form_grade_tb"><abbr title="' + (collectibleType==='coin'?'Fine':(collectibleType==='banknote'?'Fine':'Fine')) + '">' + (collectibleType==='coin'?'F':(collectibleType==='banknote'?'F':'F')) + '</abbr></label>\<input type="radio" name="collec_form_grade" value="ttb" id="collec_form_grade_ttb"' + (grade==='ttb'?' checked="checked"':'') + ' /><label for="collec_form_grade_ttb"><abbr title="' + (collectibleType==='coin'?'Very Fine':(collectibleType==='banknote'?'Very Fine':'Very Fine')) + '">' + (collectibleType==='coin'?'VF':(collectibleType==='banknote'?'VF':'VF')) + '</abbr></label>\<input type="radio" name="collec_form_grade" value="sup" id="collec_form_grade_sup"' + (grade==='sup'?' checked="checked"':'') + ' /><label for="collec_form_grade_sup"><abbr title="' + (collectibleType==='coin'?'Extremely Fine':(collectibleType==='banknote'?'Extremely Fine':'Extremely Fine')) + '">' + (collectibleType==='coin'?'XF':(collectibleType==='banknote'?'XF':'XF')) + '</abbr></label>\<input type="radio" name="collec_form_grade" value="spl" id="collec_form_grade_spl"' + (grade==='spl'?' checked="checked"':'') + ' /><label for="collec_form_grade_spl"><abbr title="' + (collectibleType==='coin'?'Almost Uncirculated':(collectibleType==='banknote'?'Almost Uncirculated':'Almost Uncirculated')) + '">' + (collectibleType==='coin'?'AU':(collectibleType==='banknote'?'AU':'AU')) + '</abbr></label>\<input type="radio" name="collec_form_grade" value="fdc" id="collec_form_grade_fdc"' + (grade==='fdc'?' checked="checked"':'') + ' /><label for="collec_form_grade_fdc"><abbr title="' + (collectibleType==='coin'?'Uncirculated':(collectibleType==='banknote'?'Uncirculated':'Uncirculated')) + '">' + (collectibleType==='coin'?'UNC':(collectibleType==='banknote'?'UNC':'UNC')) + '</abbr></label>\                    </span>\
            </div>\
            \
            <div>\
                <label class="form_name" for="collec_form_quantity">Quantity:</label>\
                <span class="form_value"><input type="number" name="collec_form_quantity" id="collec_form_quantity" min="1" value="' + (quantity==null?'1':quantity) + '" /></span>\
            </div>\
            \
            <div>\
                <label class="form_name" for="collec_form_value">Buying value (USD):</label>\
                <span class="form_value"><input type="number" name="collec_form_value" id="collec_form_value" step="0.01" min="0"' + (value==null?'':' value="'+value+'"') + ' /></span>\
            </div>\
            \
            <div>\
                <label class="form_name" for="collec_form_comment">Private comment:</label>\
                <span class="form_value"><input type="text" name="collec_form_comment" id="collec_form_comment" value="' + (comment==null?'':htmlEncode(comment)) + '" /></span>\
            </div>\
            \
            <div>\
                <span class="form_name">For exchange:</span>\
                <span class="form_value">\
                    <input type="radio" name="collec_form_swap" value="1" id="collec_form_swap_yes"' + (forSwap==1?' checked="checked"':'') + ' /><label for="collec_form_swap_yes">yes</label>\
                    <input type="radio" name="collec_form_swap" value="0" id="collec_form_swap_no"' + (forSwap!=1?' checked="checked"':'') + ' /><label for="collec_form_swap_no">no</label>\
                </span>\
            </div>\
            \
            <div>\
                <label class="form_name" for="collec_form_swap_comment">Public comment:</label>\
                <span class="form_value"><input type="text" name="collec_form_swap_comment" id="collec_form_swap_comment" value="' + (swapComment==null?'':htmlEncode(swapComment)) + '" /></span>\
            </div>\
            \
                            <div>\
                <label class="form_name" for="collec_form_section">Collection:</label>\
                <span class="form_value">\
                    <select name="collec_form_section" id="collec_form_section">\
                        <option value="43709"' + (section==43709?' selected="selected"':'') + '>Misc - Non-US</option><option value="43708"' + (section==43708?' selected="selected"':'') + '>Misc - US</option><option value="43706"' + (section==43706?' selected="selected"':'') + '>Type - France</option><option value="43705"' + (section==43705?' selected="selected"':'') + '>Type - Mexico</option><option value="43707"' + (section==43707?' selected="selected"':'') + '>Type - US</option>\
                    </select>\
                </span>\
            </div>\
                            \
            <div>\
                <label class="collec_form_pictures" for="collec_form_pictures">Pictures or<br />private PDF documents:</label>\
                <span class="form_value">\
                    <div style="display:none;"><select multiple="multiple" name="collec_form_pictures" id="collec_form_pictures">';
    if (pictures) {
        for (var i=0; i<pictures.length; i++) {
            content = content + '<option value="' + pictures[i] +  '" selected="selected">' + pictures[i] +  '</option>';
        }
    }
    content = content + '</select></div>\
                    <div id="collec_form_dropzone" class="dropzone" style="display:block; position:relative"></div>\
                </span>\
            </div>\
            \
        </div>\
        <div class="collec_form_submit">\
            <input type="submit" class="button" value="Save" onclick="save_collection(false); return false;">';
    if(action=='create') content = content +'\
            <input type="button" class="button secondary" value="Save and add again" onclick="save_collection(true); return false;">';
    content = content +'\
        </div>\
    </form></div>';
    modal.innerHTML = content;
    $('body').append(modal);

    let myDropzone = new Dropzone('#collec_form_dropzone', {
        url: '/vous/upload_picture.php', 
        acceptedFiles: '.jpeg,.jpg,.png,.gif,.pdf',
        addRemoveLinks: true,
        dictDefaultMessage: 'Click or drop files here to upload.',
        dictInvalidFileType: 'Only JPEG, PNG, GIF and PDF files are accepted.',
        dictCancelUpload: 'Cancel upload',
        dictUploadCanceled: 'Upload canceled.',
        dictCancelUploadConfirmation: 'Are you sure you want to cancel this upload?',
        dictRemoveFile: 'Remove',
        dictRemoveFileConfirmation: 'Are you sure you want to remove this file?'});
    myDropzone.on("success", function(file, response) {
        if (response.substr(0,2)!=='OK') {
            alert(response);
            myDropzone.removeFile(file);
        }
        else {
            let filename = response.substr(2);
            file.serverFileName = filename;
            $('#collec_form_pictures').append($('<option>', { value: filename, text: filename, selected: true }));
            if (filename.endsWith('.pdf')) {
                let thumbnail_url = filename.replace('.pdf','-360.png');
                this.emit('thumbnail', file, "https://en.numista.com/vous/pictures/237078/" + thumbnail_url);
            }
        }
    });
    myDropzone.on("removedfile", function(file) {
        if (file.serverFileName) {
            $('#collec_form_pictures option[value="'+file.serverFileName+'"]').prop('selected', false);
        }
    });
    myDropzone.on("addedfile", function(file) {
        $('.collec_form_submit input').prop('disabled', true);
    });
    myDropzone.on("queuecomplete", function() {
        $('.collec_form_submit input').prop('disabled', false);
    });
    if (pictures) {
        for (let i=0; i<pictures.length; i++) {
            let mockFile = { name: pictures[i], serverFileName: pictures[i] };
            let thumbnail_url = pictures[i].endsWith('.pdf')?pictures[i].replace('.pdf','-360.png'):pictures[i].replace('.','-360.');
            myDropzone.displayExistingFile(mockFile, "/vous/pictures/237078/"+thumbnail_url);
            myDropzone.files.push(mockFile);
        }
    }

    let checked_grade = $('input[name=collec_form_grade]:checked').val();
    $('input[name=collec_form_grade]').on("click", function() {
        if($(this).val() === checked_grade){
            $(this).prop("checked",false);
            checked_grade = null;
        }
        else {
            checked_grade = $(this).val();
        }
    });
    
    $('#collec_form_quantity').focus();
    $(document).keyup(function(e) {
        if (e.key === "Escape") collec_modal_close();
    });
}


function veux(checkbox) {
    let type = checkbox.className.replace('check_veux', '');
    let veux = $(checkbox).prop('checked');
    
    $('.check_veux'+type).prop('disabled', true);
    $('.check_veux_pas'+type).prop('disabled', true);
    $('.check_veux'+type).prop('checked', veux);
    $('.check_veux_pas'+type).prop('checked', false);
    
    $.get("/souhaits/modifier_veux.php", {type: type, veux: veux, veux_pas: 'false'})
        .done(function(reponse) {
            if (reponse!=="OK") alert("An error occurred. Please try again.");
            $('.check_veux'+type).prop('disabled', false);
            $('.check_veux_pas'+type).prop('disabled', false);
        });
}

function veux_pas(checkbox) {
    let type = checkbox.className.replace('check_veux_pas', '');
    let veux_pas = $(checkbox).prop('checked');
    
    $('.check_veux'+type).prop('disabled', true);
    $('.check_veux_pas'+type).prop('disabled', true);
    $('.check_veux'+type).prop('checked', false);
    $('.check_veux_pas'+type).prop('checked', veux_pas);
    
    $.get("/souhaits/modifier_veux.php", {type: type, veux: 'false', veux_pas: veux_pas})
        .done(function(reponse) {
            if (reponse!=="OK") alert("An error occurred. Please try again.");
            $('.check_veux'+type).prop('disabled', false);
            $('.check_veux_pas'+type).prop('disabled', false);
        });
}

function modifier(piece, etat, type_piece) {
    $("#"+etat+piece).css("background", "#FF99CC");
    let quantite = $("#"+etat+piece).val();
    
    $.get("/vous/save_collection_old_style.php", {piece: piece, etat: etat, quantite: quantite})
        .done(function(reponse) {
            if (reponse==="OK") {
                $("#"+etat+piece).css("background", "#FFFFFF");
                document.getElementById(etat+piece).style.background = "#FFFFFF";
                if (typeof type_piece!="undefined") {
                    let compte_collec = 0;
                    let compte_ech = 0;
                    let champs = document.getElementsByTagName("input");
                    let n = champs.length;
                    for (let i=0; i < n; i++) {
                        if (champs[i].className==="qc"+type_piece && parseInt(champs[i].value))
                            compte_collec += parseInt(champs[i].value);
                        else if (champs[i].className==="qe"+type_piece && parseInt(champs[i].value))
                            compte_ech += parseInt(champs[i].value);
                    }
                    self.document.getElementById("affichage_qc"+type_piece).innerHTML = compte_collec;
                    let display = compte_ech===0? 'none' : 'inline';
                    self.document.getElementById("affichage_qe_cadre"+type_piece).style.display = display;
                    self.document.getElementById("affichage_qe"+type_piece).innerHTML = compte_ech;
                }
            }
            else alert("An error occurred. Please try again.");
        });
}

function comm_perso(piece) {
    let ancienCommentaire = $('#comm'+piece).text();
    let nouveauCommentaire = prompt("Enter your personal comment.\nYou may register varieties, errors...", ancienCommentaire);
    if (nouveauCommentaire!=null) {
        $.get("/vous/save_comment.php", {piece: piece, texte: nouveauCommentaire})
            .done(function(reponse) {
                if (reponse==="OK") {
                    $("#comm"+piece).html(htmlEncode(nouveauCommentaire));
                    $("#lien_comm"+piece)[0].onclick = function () { comm_perso(piece, nouveauCommentaire); }
                }
                else if (reponse==='NO_ITEMS') {
                    // pass
                }
                else alert("An error occurred. Please try again.");
            });
    }
}
